### ISSUE TITLE
Wrong type narrowing when types are compatibles

### ISSUE THREAD

**Issue by akarzazi** [Microsoft/TypeScript#11664]
**TypeScript Version:**  2.0.3 

**Code**

``` ts
class Result1 {
    readonly value1: number;

    constructor(value1: number) {
        this.value1 = value1;
    }
}

class Result2 {
    readonly value1: number;
    readonly value2: number;

    constructor(value1: number, value2: number) {
        this.value1 = value1;
        this.value2 = value2;
    }
}

function someFunction(result: Result1 | Result2) {
    if (result instanceof Result1) {
        // result is still Result1 | Result2
        console.log("1");
    }
    else {
        // result is never ! this is wrong !
        console.log("2");
    }
}

// this will output 2 !
someFunction(new Result2(9,9))
```

**Expected behavior:**
instance of type guard should not care about type compatibility
**Actual behavior:** 
Wrong assertion by the compiler.

**Comment by yortus** [comment]
Essentially a duplicate of #10934. See also design notes about `instanceof` in #8503.

**Comment by akarzazi** [comment]
Indeed, this is a duplicate :(
From #7271, the issue will be fixed in 2.1.

**Comment by yortus** [comment]
@akarzazi note that the [fix](https://github.com/Microsoft/TypeScript/pull/10216) for #7271 only affects narrowing structurally *identical* types, but your issue here is with narrowing structural subtype/supertypes. This hasn't been solved yet and is on the roadmap for [future](https://github.com/Microsoft/TypeScript/wiki/Roadmap#future) investigation at an unspecified time.

**Comment by akarzazi** [comment]
@yortus I thought the fix will cover type compatibly as well.
Since mhegady said : As noted in #8503, classes should be treated definitely for instanceof checks.

**Comment by mikelehen** [comment]
FYI- This still reproduces with TypeScript 2.3.2

**Comment by nodefish** [comment]
Really simplistic example of what's wrong with instanceof:

```javascript
const model = this.opts.resources.model.store.get(id);
if (!model || model instanceof Placeholder)
    return null;
const data: any = model.get();
```

TSC gives the following error:
`Property 'get' does not exist on type 'never'.`

A workaround was proposed in #13325 but I can't see an equivalent in imperative code sections.

**Comment by RyanCavanaugh** [comment]
Fixed in #19671
